sig-scalability/slos/pod_startup_latency.md

```diff
@@ -0,0 +1,47 @@
+## Pod startup latency SLI/SLO details
+
+### User stories
+- As a user of vanilla Kubernetes, I want some guarantee how quickly my pods
+will be started.
+
+### Other notes
+- Only schedulable and stateless pods contribute to the SLI:
+  - If there is no space in the cluster to place the pod, there is not much
+    we can do about it (it is task for Cluster Autoscaler which should have
+    separate SLIs/SLOs).
+  - If placing a pod requires preempting other pods, that may heavily depend
+    on the application (e.g. on their graceful termination period). We don't
+    want that to contribute to this SLI.
+  - Mounting disks required by non-stateless pods may potentially also require
+    non-negligible time, not fully dependent on Kubernetes.
+- We are explicitly excluding image pulling from time the SLI. This is
+because it highly depends on locality of the image, image registry performance
+characteristic (e.g. throughput), image size itself, etc. Since we have
+no control over any of those (and all of those would significantly affect SLI)
+we decided to simply exclude it.
+- We are also explicitly excluding time to run init containers, as, again, this
+is heavily application-dependent (and does't depend on Kubernetes itself).
+- The answer to question "when pod should be considered as started" is also
+not obvious. We decided for the semantic of "when all its containers are
+reported as running and observed via watch", because:
+  - we don't want to rely on "readiness checks", because they heavily
+    depend on the application. If the application takes couple minutes to
+    initialize before it starts responding to readiness checks, that shouldn't
+		count towards Kubernetes performance.
+  - even if your application started, many control loops in Kubernetes will
+    not fire before they will observe that. If Kubelet is not able to report
+    the status due to some reason, other parts of the system will not have
+    a way to learn about it - this is why reporting part is so important
+    here.
+  - since watch is so centric to Kubernetes (and many control loops are
```

Comment:
I'm sure this was a conscious decision in the past and not asking to change this, but one might also reason why watch latency shouldn't be included. For e.g a client who's listing directly from etcd or even from watch cache (where the pod is updated but the corresponding event wasn't delivered yet) may not suffer from this latency :)

Reply:
Agree - but if we're talking about core components (e.g. endpoints controller), it watches.

Reply:
Sure.. I'm just saying that there are users whose view of 'pod startup' doesn't have this latency. And controllers are typically not user-facing - wdyt?

Reply:
> I'm just saying that there are users whose view of 'pod startup' doesn't have this latency. 

Yes

> And controllers are typically not user-facing - wdyt?

But if your pod is part of service, what you care in the end is how fast it will be added to your service.
So this generally matters (if your pods is not part of service, in many cases you don't even care about its status being reported).

That said, yes - it's debatable whether we want watch or no. Added a TODO to revisit that.

Reply:
That lgtm - thnx.

Reply:
Btw - I think the part you mention about pod being available in a service sounds like a different SLI to me (which of course would be influenced by watch latency - and we anyway seem to have another SLI for watch latency) :)

Reply:
Yes - being available in a service is a different SLI.
But it cannot even become part of service until kubelet reports its status (and that status will be observed) - only then something may start processing.
I can imagine an artificial scenario that "pod status update latency" have 10x higher watch latency. Should this then contribute to service SLI - I think no. This will ideally be a separate SLI, but since we're not there yet, I think it's better to have it here...
But yeah - TODO is already there.

Reply:
Yes, and that is why I was mentioning "we anyway seem to have another SLI for watch latency" for those kind of situations (which is an SLI we're already having (at least in our docs)?).
Anyway - I raised that just for discussion.